src/libplctag/Generic/GenericTag.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+
+namespace libplctag.Generic
+{
+    public class GenericTag<TPlcType, TDotNetType> : ITag, IGenericTag<TDotNetType> where TPlcType : IPlcType<TDotNetType>, new()
+    {
+
+        private readonly TPlcType plcType;
+        private readonly Tag tag;
+
+        /// <summary>
+        /// Provides a new tag. If the CPU type is LGX, the port type and slot has to be specified.
+        /// </summary>
+        /// <param name="gateway">IP address of the gateway for this protocol. Could be the IP address of the PLC you want to access.</param>
+        /// <param name="path">Required for LGX, Optional for PLC/SLC/MLGX IOI path to access the PLC from the gateway.
+        /// <param name="cpuType">Allen-Bradley CPU model</param>
+        /// <param name="name">The textual name of the tag to access. The name is anything allowed by the protocol. E.g. myDataStruct.rotationTimer.ACC, myDINTArray[42] etc.</param>
+        /// <param name="elementCount">elements count: 1- single, n-array.</param>
+        /// <param name="millisecondTimeout"></param>
+        /// <param name="debugLevel"></param>
+        /// <param name="protocol">Currently only ab_eip supported.</param>
+        /// <param name="readCacheMillisecondDuration">Set the amount of time to cache read results</param>
+        /// <param name="useConnectedMessaging">Control whether to use connected or unconnected messaging.</param>
+        public GenericTag(IPAddress gateway,
+                   string path,
+                   CpuType cpuType,
+                   string name,
+                   int millisecondTimeout,
+                   int elementCount = 1,
+                   DebugLevel debugLevel = DebugLevel.None,
+                   Protocol protocol = Protocol.ab_eip,
+                   int readCacheMillisecondDuration = default,
+                   bool useConnectedMessaging = true)
+        {
+            //Instantiate our definition
+            //TODO: These could be singleton or a private static lookup for performance
+
+            plcType = new TPlcType();
+            var elementSize = plcType.ElementSize;
+
+
+            this.tag = new Tag(gateway,
+                               path,
+                               cpuType,
+                               elementSize,
+                               name,
+                               millisecondTimeout,
+                               elementCount,
+                               debugLevel,
+                               protocol,
+                               readCacheMillisecondDuration,
+                               useConnectedMessaging);
+        }
+
+        public TDotNetType Value { get => plcType.Decode(tag); set => plcType.Encode(tag, value); }
```

Comment:
> To supports arrays could we expose this as an indexer instead of a property, and pass the index to the encode/decode functions?

I think something like that will work. I've never had a reason to dig into custom indexers but I think this is a good fit. I've got some stashed changes that were starting to implement Arrays. I'll start a branch after I merge this.

I reverted my commit that changed the package refs around. I realized my mistake after I did that, but wanted to step into the low level code during debug so I left it while developing.